Milestone 2

adriaanbd opened the pull request:
Implementation for the following requirements:
1. Prepared the directory structure (created components, containers, actions and reducers directory)
2. Implemented App component (wrapping BooksList and BooksForm)
3. Implemented BooksList component (displays books as html table and fetches books from store)
4. Implemented Book component (accepts book obj as prop and represents single row in books table)
5. Implemented BooksForm component (very simple form for adding books with input, select box with 7 categories and submit button)
6. Defined Redux actions and reducers (create book and remove book action creators with reducers)
7. Initialized the Redux store (creates initial state, renders app component within providers with store)

Please note that I prefer the functional component approach to React, as opposed to class based components. Futhermore, I've implemented the useSelector hook instead of mapStateToProps, and will be implementing useDispatch hook as well.

Moreover, I've disabled two eslint rules with justification:
* `"jsx-a11y/label-has-associated-control": "off"` because I have indeed defined an associated control for the select box label in the form of `label htmlFor="categories"` and `select id="categories"`
* `"jsx-a11y/label-has-associated-control": "off"` because this is deprecated in favor of the one above, read [here](https://github.com/evcohen/eslint-plugin-jsx-a11y/blob/master/docs/rules/label-has-for.md)
* `"react/no-array-index-key": "off"` because the options constant in the select box do not have a natural ID beyond it's array index, this rule is due to the fact that the IDs might change, but in this case they are constant. This rule is pertaining to: `{ CATEGORIES.map((cat, idx) => <option key={idx} value={cat}>{cat}</option>) }`. We need to insert a `key` and the [documentation](https://reactjs.org/docs/lists-and-keys.html#keys) says it's fine to insert index in this case, specifically it states: 
> When you don’t have stable IDs for rendered items, you may use the item index as a key as a last resort.

I'm authorized to work alone in this project. Looking forward to your review.

adriaanbd committed:
implemented rough app component

adriaanbd committed:
implemented rough Book and BooksList component

adriaanbd committed:
added prop-types dependency

adriaanbd committed:
added tbody

adriaanbd committed:
implemented rough BooksForm component

adriaanbd committed:
updated BooksList to include useSelector within component

adriaanbd committed:
moved booklist to container

adriaanbd committed:
excluding rules because of justified exception

adriaanbd committed:
adding store, reducers and actions

adriaanbd committed:
fixed nested label in select issue

adriaanbd committed:
correctly destructure books passing props to book

adriaanbd committed:
moving app to components

adriaanbd committed:
wrapping app in provider to variable

adriaanbd committed:
fixing single param paren errors

adriaanbd committed:
disabling jxs label has for due bc its deprecated

shloch commented on src/containers/BooksList.js:
```diff
@@ -0,0 +1,36 @@
+  );
+}
+
+export default BooksList;
```
You forgot the following
- connect to redux store ([see requirements](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/react_reducx/bookstore/milestone2/BooksList/ConnetToRedux.png))

- use `mapStateToProps` to fetch books from the `store`

adriaanbd replied:
I have implemented those, in the form of hooks:
* `useSelector`, please read more about it [here](https://react-redux.js.org/next/api/hooks). Here is a quick explanation in the documentation:
> The selector is approximately equivalent to the mapStateToProps argument to connect conceptually. The selector will be called with the entire Redux store state as its only argument. The selector will be run whenever the function component renders. useSelector() will also subscribe to the Redux store, and run your selector whenever an action is dispatched.

This requirement has been met in the form of hooks, please re-consider.

shloch replied:
thanks for enlightenment...i will definitely learn hooks today from your code  😊

shloch replied:
just correct, what's to be corrected and re-file for review. We'll re-consult the project and taking hooks into consideration 😊

shloch commented on src/containers/BooksList.js:
```diff
@@ -0,0 +1,36 @@
+            title={book.title}
+            category={book.category}
+          />
+        )) }
```
- pass the `book` object as a prop to each `Book` component  (pass it as an OBJECT, not split as you did). Do something like :
``` 
 <Book key={index} book={book} />
```

shloch commented on src/components/BooksForm.js:
```diff
@@ -0,0 +1,38 @@
+  );
+}
+
+export default BooksForm;
```
You forgot the following :
- it has a state with `title` and `category` 
- this file (`BookForm.js`), should be in `src/containers/` folder,  it's a "container" into which books will be filled , so relocate the fle 😊

adriaanbd replied:
Oh that's right, BookForm will indeed become a container later on because it will communicate with the redux store and also manage its own state, is that the reason why it should be a container? My understanding was that if the components gets or sets state it is a container component. In other words, if it only generates HTML it is a presentational component.

adriaanbd replied:
In regards to BookForm having its own state, I've definitely included that for Milestone 3, however despite the guidelines requiring this item in this Milestone, it isn't required in Milestone 2 on pathwright:
![image](https://user-images.githubusercontent.com/22228397/73593604-ff493f00-44d3-11ea-9be0-b2377ea93de7.png)

shloch replied:
oooh...concerning the file relocation, it could be anywhere..... just that it's on the requirements for it to be here (src/containers/ ), no deeper explanation to that 😊😊

![Screen Shot 2020-02-01 at 3 36 44 PM](https://user-images.githubusercontent.com/40712459/73593860-a9db5500-4508-11ea-9807-944b1390809d.png)

shloch reviewed (changes requested):
## Hello @adriaanbd 
Great job there, from your code, you perfectly understand the requirements
There's few steps/things to do though to meet up with the requirements. You are on the right track, just some minor completions to do 😊😊

See comments below for **required  changes** to do

adriaanbd committed:
moving BookForm to container dir

adriaanbd committed:
passing book as obj to props

adriaanbd committed:
disabling forbidden obj prop type exception

adriaanbd commented:
Addressed changes:
* Passing book object as props to Book component
* Moved BookForm component to container directory

Please note that:
* This project favors functional components and uses Hooks to connect to react state and redux store.
* Pathwright requirements do not require BookForm to have state yet (this is required in Milestone 3, not Milestone 2)
* I've disabled  `"react/forbid-prop-types": "off"` rule due to `book: PropTypes.object.isRequired` in Book component because I need to pass an object as props to Book component and the rule [documenation](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/forbid-prop-types.md) states the following:
> This rule is a formatting/documenting preference and not following it won't negatively affect the quality of your code. This rule encourages prop types that more specifically document their usage.

Furthermore, please take into account previous review comments and conversation.

Looking forward to your review.

shloch commented on src/containers/BooksList.js:
```diff
@@ -0,0 +1,34 @@
+import React from 'react';
+import { useSelector } from 'react-redux';
```
thanks for the hooks approach eye-opener, i appreciate.....learnt from your project 😊

shloch commented on src/containers/BooksList.js:
```diff
@@ -0,0 +1,34 @@
+      </thead>
+      <tbody>
+        { books.map(book => (
+          <Book
+            key={book.bookId}
+            book={book}
```
👏👏👏

shloch reviewed (approved):
## Hello @adriaanbd 
Great job......👏👏
![accomplished](https://user-images.githubusercontent.com/40712459/73598134-6696db80-4534-11ea-9972-91948afbb005.jpg)

### You can merge this
Happy coding 😊 😊
